modules/generator/registry/native_histogram.go

```diff
@@ -0,0 +1,379 @@
+
+	newSeries := h.newSeries(labelValueCombo, value, traceID, multiplier)
+	h.series[hash] = newSeries
+}
+
+func (h *nativeHistogram) newSeries(labelValueCombo *LabelValueCombo, value float64, traceID string, multiplier float64) *nativeHistogramSeries {
+	newSeries := &nativeHistogramSeries{
+		// TODO move these labels in HistogramOpts.ConstLabels?
+		labels: labelValueCombo.getLabelPair(),
+		promHistogram: prometheus.NewHistogram(prometheus.HistogramOpts{
+			Name:    h.name(),
+			Help:    "Native histogram for metric " + h.name(),
+			Buckets: h.buckets,
+			// TODO check if these values are sensible and break them out
+			NativeHistogramBucketFactor:     1.1,
+			NativeHistogramMaxBucketNumber:  100,
+			NativeHistogramMinResetDuration: 15 * time.Minute,
+		}),
+		lastUpdated: 0,
+	}
+
+	h.updateSeries(newSeries, value, traceID, multiplier)
+
+	return newSeries
+}
+
+func (h *nativeHistogram) updateSeries(s *nativeHistogramSeries, value float64, traceID string, multiplier float64) {
+	for i := 0.0; i < multiplier; i++ {
+		s.promHistogram.(prometheus.ExemplarObserver).ObserveWithExemplar(
+			value,
+			map[string]string{h.traceIDLabelName: traceID},
+		)
+	}
+	s.lastUpdated = time.Now().UnixMilli()
+}
+
+func (h *nativeHistogram) name() string {
+	return h.metricName
+}
+
+func (h *nativeHistogram) collectMetrics(appender storage.Appender, timeMs int64, externalLabels map[string]string) (activeSeries int, err error) {
+	h.seriesMtx.Lock()
+	defer h.seriesMtx.Unlock()
+
+	lbls := make(labels.Labels, 1+len(externalLabels))
+	lb := labels.NewBuilder(lbls)
+	activeSeries = 0
+
+	lb.Set(labels.MetricName, h.metricName)
+
+	// set external labels
+	for name, value := range externalLabels {
+		lb.Set(name, value)
+	}
+
+	for _, s := range h.series {
+
+		// Set series-specific labels
+		for i, name := range s.labels.names {
+			lb.Set(name, s.labels.values[i])
+		}
+
+		// Extract histogram
+		encodedMetric := &dto.Metric{}
+
+		// Encode to protobuf representation
+		err = s.promHistogram.Write(encodedMetric)
+		if err != nil {
+			return activeSeries, err
+		}
+
+		// NOTE: Store the encoded histogram here so we can keep track of the exemplars
+		// that have been sent.  The value is updated here, but the pointers remain
+		// the same, and so Reset() call below can be used to clear the exemplars.
+		s.histogram = encodedMetric.GetHistogram()
+
+		// NOTE: A subtle point here is that the new histogram implementation is
+		// chosen in the registry.  This means that if the user overrides change
+		// from "both" to "classic", we will still be using the new histograms
+		// implementation. This is because the registry is not recreated when the
+		// overrides change. This is a tradeoff to avoid recreating the registry
+		// for every change in the overrides, but it does mean that even if user
+		// changes to "classic" histograms, we will still be using the new
+		// histograms implementation and want to avoid generating native
+		// histograms.
```

Comment:
This will work when we switch from `native` or `both` to `classic`, but it will not work when we go from `classic` to anything else. If we initiate their tenant with `classic` we use the old histogram implementation, when we change the override nothing will be updated and we keep using this implementation.
To workaround this we need to recreate the processors I think so they all re-register their histograms.

But maybe something quick to fix in a follow-up PR?

Reply:
Right, I think let's follow up.